Implement Transforms

alicanb opened the pull request:
This implements a class hierarchy of `Transform`s, a `TransformedDistribution` class, and `LogNormal` and `HalfNormal` classes as examples of `TransformedDistribution`s. `Transform` objects have `.forward()` and `.inverse()` methods. Most transforms are invertible and also have `.log_abs_det_jacobian()` methods. The forward and inverse methods are bidirectionally memoized so that e.g. when `.forward()` is called, subsequent calls to both `.forward()` and `.inverse()` are free.

Not all transforms are invertible, e.g. `AbsTransform` is a 2-to-1 mapping, and `LogprobTransform` is projects an entire dimension away. However both of these have `.inverse()` implemented as pseudoinverses. cc @apaszke

alicanb committed:
add bijectors

alicanb committed:
Merge branch 'master' of github.com:pytorch/pytorch into bijector

alicanb committed:
add entropy

alicanb committed:
allow chaining of bijectors

alicanb committed:
add affinebijector

alicanb committed:
use log_abs_det_jacobian

fritzo committed:
Changed to cache-by-default; add x,y args to log_deg_jacobian method

fritzo committed:
Fix TransformedDistribution.rsample

fritzo committed:
Add KL divergence for LogNormal distribution

fritzo committed:
Add .domain, .codomain constraints to Bijectors; add tests

fritzo committed:
Add .support property to TransformedDistribution

fritzo committed:
Add AffineBijector

fritzo committed:
Add SigmoidBijector

alicanb committed:
change bijectors to transforms, add half-normal

fritzo committed:
Rename-refactor test_distributions.py

alicanb committed:
remove test_transformed_distributions

alicanb committed:
fix halfnormal docs

alicanb committed:
remove log_abs_det_jacobian

alicanb committed:
add transforms to distributions.__all__

fritzo committed:
Add Loprob and StickBreaking transforms

fritzo committed:
Add docs for transforms

fritzo committed:
Merge branch 'master' into bijector

yf225 commented:
@pytorchbot retest this please

apaszke commented on torch/distributions/transforms.py:
```diff
@@ -0,0 +1,231 @@
+    def __init__(self):
+        self._cache = {}
+
+    def forward(self, x):
```
forward/backward terms appearing in PyTorch are mostly from backpropagation/reverse-mode AD, and I don't think that this word is often used in context of transforms/bijections/diffeomorphisms. I think using `__call__` would be a nicer API overall. Then, you'd have:
```python
f = SomeTransform()
f_inv = f.inverse

f_inv(f(x)) == x
```
---
Another possibility would be to make `.inverse` a property and  instantiate an `InverseTransform(self)` inside it. Then `f.inverse` would be a first-class transform too, and you could do things like `f.inverse.log_abs_det_jacobian(a, b)` (if this pattern ends up being used often it might be good to shorten `inverse` to `inv`). Also `f.inv.inv is f`

apaszke commented on torch/distributions/transforms.py:
```diff
@@ -0,0 +1,231 @@
+        shape = x.shape
+        for _ in range(self.event_dim):
+            result = result.sum(-1)
+            shape = shape[:-1]
```
This would be faster:
```python
result = torch.abs(self.scale).log()
shape = x.shape
if self.event_dim:
    # NOTE: no need for contiguous here
    result = result.view(*result.size()[:-self.event_dim], -1).sum(-1)
    shape = shape[:-self.event_dim]
return results.expand(shape)
```

alicanb replied:
Python2 complains about `result.view(*result.size()[:-self.event_dim], -1)`

apaszke commented on torch/distributions/transforms.py:
```diff
@@ -0,0 +1,231 @@
+class StickBreakingTransform(Transform):
+    """
+    Transform from the simplex to unconstrained of one fewer dimension via a
+    stick-breaking process.
```
Some explanation of stick-breaking (possibly a link) would be nice (it's not trivial like all other transforms)

fritzo replied:
Done.

apaszke commented on torch/distributions/transforms.py:
```diff
@@ -0,0 +1,231 @@
+    """
+    Transform from the simplex to unconstrained space via `y = log(x)`.
+
+    This is not bijective and cannot be used for HMC. However this acts mostly
```
Should we have a common super-class/flag for bijective transforms? Currently there's no way to tell other than looking in the docs.

fritzo replied:
Done. (added a `.bijective` class attribute)

apaszke reviewed (commented):
Looks good, but I'd like to discuss a few things before merging this.

alicanb committed:
address low-hanging comments

alicanb commented on torch/distributions/transforms.py:
```diff
@@ -167,6 +174,9 @@ def __init__(self, loc, scale, event_dim=0):
         self.scale = scale
         self.event_dim = event_dim
 
+    def __eq__(self, other):
```
I'm pretty sure this is not correct. Should this return a bool matrix for batched `TransformedDistribution`s?

apaszke replied:
No, `__eq__` should always return a Python boolean

alicanb commented:
I fixed the low-hanging comments. I also realized we forgot to add tests for `LogNormal` and `HalfNormal`, so I added them as well. I think scipy's `lognorm.entropy()` [link](https://github.com/scipy/scipy/blob/11509c4a98edded6c59423ac44ca1b7f28fba1fd/scipy/stats/_continuous_distns.py#L3641) is incorrect. so I omitted that test (it passes the monte carlo test).

apaszke commented on torch/distributions/transforms.py:
```diff
@@ -0,0 +1,242 @@
+        self._cache = {}
+
+    def __eq__(self, other):
+        return type(other) is type(self)
```
That's probably a bad idea. Think what happens when someone implements a transform that is somehow parametrized, but forgets to implement `__eq__`

alicanb replied:
We may do something like this: introduce a flag for non-parameterized transforms, default `__eq__` to check for types if `self` and `other` are both non-parameterized, otherwise raise `NotImplementedError`
 
For parameterized transforms, I have no idea how to handle batches.

fritzo replied:
We don't need to handle batches; that sounds a bit over-engineered. In fact I think it's safe to define `Transform.__eq__` by `self is other` by default and define `type(self) == type(other)` for the each of few parameter-free transforms. WDYT?

apaszke commented on torch/distributions/transforms.py:
```diff
@@ -0,0 +1,242 @@
+        if self.event_dim:
+            # NOTE: no need for contiguous here
+            result = result.view(*result.size()[:-self.event_dim], -1).sum(-1)
+            shape = shape[:-self.event_dim]
```
To fix Python 2 compatibility you can do that:
```
result_size = result.size()[:-self.event_dim] + (-1,)
result = result.view(result_size).sum(-1)
```
Also, please remove the note (it was just a general observation, but doesn't need to be in the code)

alicanb committed:
keep python2 happy

fritzo commented on torch/distributions/transforms.py:
```diff
@@ -0,0 +1,231 @@
+    """
+
+    def __init__(self):
+        self._cache = {}
```
I'm open to ideas here. This seemed like the least weird implementation I could find.

The purposes are (1) to speed up inverses, (2) to compute inverses in cases where they have not been implemented, and (3) cache intermediate computations for computing `log_abs_det_jacobian()` to avoid gradients being ping-ponged back and forth through `.forward()` and `.inverse()` (e.g. if `.log_deg_jacobian_inverse(x, y)` depends on x, we want to avoid `x` being defined in terms of `dist(inverse(dist.forward(y))`. The Pyro implementation automatically ejects an item from cache the first time it is read, but that behavior seemed to make it too easy to introduce a leak.

We could implement cache size of 1, but that could lead to silent degradation of performance and gradient stability. Alican originally had something like a `.memoize` flag to configure this, but I removed it at some point to simplify the interface.

fritzo replied:
Just to clarify: sometimes `.inverse()` is not implemented, and sometimes it is 100x more expensive to compute than to memoize (e.g. `gesv` vs `mm`).

apaszke replied:
I think this makes sense, but I'm worried that this will lead to space leaks - users might not expect us to hold on to memory indefinitely.

fritzo replied:
Ok after design discussion with @martinjankowiak I've changed this to:
- Do no caching by default in `Transform`
- Split out `CachedTransform` that implements a LRU(1) cache
- Only a couple of the more complex transforms inherit from `CachedTransform`, but in the future many more will, e.g. `NormalizingFlow`, `NADE`, ...

fritzo commented on torch/distributions/half_normal.py:
```diff
@@ -0,0 +1,39 @@
+from torch.distributions.transformed_distribution import TransformedDistribution
+
+
+class HalfNormal(TransformedDistribution):
```
I agree this one might not be very useful. @jpchen is working on a `HalfCauchy` which is actually commonly used as a prior on variances especially in linguistics and is worth maintaining. We could use `HalfCauchy` as the canonical demonstration of `AbsTransform` and add `TruncatedNormal` in a later PR (`TruncatedNormal` is actually really tricky to get right, due to reparameterized gradients). @alicanb WDYT?

alicanb replied:
Works for me 👍 I've been thinking about how to handle `TruncatedNomal` but we can discuss that later. In the meantime, we can remove `HalfNormal` and ask @jpchen to push an implementation of `HalfCauchy` with `Transforms` to this branch (Nice thing about `HalfCauchy` is it uses both `AffineTransform` and AbsTransform`, so it will be a perfect example)

jpchen replied:
i can do that in a separate PR, i have an outstanding one in the probtorch fork which i can rewrite after this is merged, since i would need to look into the semantics of the transformations

fritzo committed:
Split CachedTransform out of Transform

fritzo committed:
Merge branch 'master' into bijector

fritzo committed:
Implement equality operators

alicanb committed:
remove HalfNormal

fritzo commented on torch/distributions/transforms.py:
```diff
@@ -0,0 +1,290 @@
+        Invokes the memoized transform `x => y`.
+        """
+        x_old, y_old = self._cached_x_y
+        if x is x_old:
```
@apaszke Is it possible in Python to the version counter of `x` to avoid subtle errors on in-place modification? E.g. could store
```py
self._cached_x_y = x, x.version, y, y.version
```

ezyang replied:
Well, if we decided we wanted to, it would be a simple matter to expose it to Python. Not sure if we want to though.

fritzo replied:
It fine to ignore `.version`. I just wanted to point out this possible use case.

apaszke replied:
I think we should verify the version. Otherwise the code can silently give incorrect results, which we're trying to avoid.

fritzo committed:
Use source order in automodule documentation

fritzo commented on torch/distributions/kl.py:
```diff
@@ -257,6 +264,13 @@ def _kl_pareto_pareto(p, q):
 
+@register_kl(TransformedDistribution, TransformedDistribution)
+def _kl_transformed_transformed(p, q):
+    if p.transforms != q.transforms:
```
I believe this has been resolved.

fritzo commented:
I'll address the `.forward()` vs `.__call__()` comment soon.

fritzo committed:
Switch from t.forward()/t.inverse() to t()/t.inv()

fritzo commented:
@apaszke `Transform` objects now support your suggested syntax `t()` and `t.inv()`. This looks very nice:
```py
t = ExpTransform()
assert isinstance(t.inv, Transform)
assert t.inv.inv is t
```

fritzo committed:
Implement LowerCholeskyTransform

alicanb commented:
I'm having a hard time with the `__call__`. I think conceptually `Transform`s should operate on `Distribution`s and return `TransformedDistribution`, ex. `AbsTransform(Normal(0, scale))` to implement `HalfNormal` (I guess `AbsTransform()(Normal(0, scale))` is close enough). Nice thing about this is if we want to do `torch.abs(Normal(0, scale))`, we can simply define `abs=AbsTransform` in transforms.py and create attribute `_torch = transforms` in `Distribution` (I think this is how it's supposed to work at least, but I tried it and it doesn't, so I may have got this completely wrong 😄)

fritzo commented:
@alicanb Here's some mathematical justification for `__call__` operating on data rather than distributions. Mathematically transforms operate pointwise as differentiable functions. When we apply these to distributions, we're really applying these to a (`.sample()`, `.log_prob()`) pair of functions bundled in a Python class. When we apply a transform to `.sample()` we're really applying a push-forward, a lifted version of the transform, to `.sample()`. When we apply a transform to `.log_prob()`, we're really applying a preimage map (or a pullback?) to the density function. Even in mathematical notation, these operations are written as `t(x)` for pointwise operation, `t_>(s)` or `t o s` for operation on a sampler, and `t^<(p)` or `p o t` for operation on a density.

fritzo committed:
Add back constraints.lower_triangular as per @tbrx request

alicanb committed:
add cache_size to init

fritzo committed:
Get cache_size argument working

fritzo committed:
Add LowerCholeskyTransform to __all__

fritzo committed:
Enforce positive definiteness in lower_cholesky constraint

fritzo committed:
Merge branch 'master' into bijector

fritzo commented:
@ezyang @apaszke I believe all that remains is to consult a the version counter when memoizing. Could you advise on how to expose that? Do you want that in this PR or in a follow-up PR (since the interface will remain unchanged)?

fritzo commented:
Looks like the build is failing due to an unrelated performance regression.

fritzo committed:
Merge branch 'master' into bijector

fritzo commented:
Looks like build is now failing due to an unrelated error in `test_squeeze_input_sizes_are_ones`.

fritzo commented:
Are there any other issues we need to address before merging this?

alicanb commented:
Are we going to do version counter checking in this PR?

fritzo commented:
> version counter checking in this PR?

I see no need to add version checking this PR, but it is @apaszke's call. Version checking seems like a small low-level change that will not affect the interface and will only be an issue in rare edge cases. Moreover this PR is blocking further work on distributions: `LogNormal`, and `ConstraintRegistry` in https://github.com/probtorch/pytorch/pull/108.

fritzo committed:
Merge branch 'master' into bijector

fritzo committed:
Fix bugs in lower triangular checks

apaszke commented on torch/distributions/transforms.py:
```diff
@@ -0,0 +1,342 @@
+        """
+        return self._call(x)
+
+    def inverse(self, y):
```
is there any reason to keep this method when you can just do `dist.inv(x)`? They should be equivalent

fritzo replied:
Good point. I was able to make this private as `._inv_call()`, but it is still needed to implement `.inv.__call__()`. We want a single object to manage both the `.__call__()` cache and the `.inv.__call__()` cache so that calling one populates the other. Thus `_Inverse` delegates caching to its `.inv` transform.

apaszke commented on torch/distributions/transforms.py:
```diff
@@ -0,0 +1,342 @@
+    """
+    Inverts a single :class:`Transform`.
+    """
+    __slots__ = ['inv']
```
Are slots necessary here? The downside is that users won't be able to assign custom attributes to transforms (not sure how important that is, but it might turn out to be useful at some point). I don't think allocating a `__dict__` would make a big difference. Also, these transforms won't support weak refs

fritzo replied:
Thanks, I didn't know about these disadvantages. Removed.

apaszke commented on torch/distributions/transforms.py:
```diff
@@ -0,0 +1,342 @@
+        return self.inv.inverse(x)
+
+    def _inverse(self, y):
+        return self.inv.__call__(y)
```
Shouldn't this call `_call` (since `__call__` might do some extra things like caching)?

fritzo replied:
As per your comment above, I was able to remove `InverseTransform._call` and `InverseTransform._inverse` in favor of the single delegating `InverseTransform.__call__`. Because this is different behavior than other `Transform`s, I've also made `_InverseTransform` private to prevent a user from accidentally creating `InverseTransform(InverseTransform(t))`. All inversion should now happen through the `.inv` property.

apaszke reviewed (changes requested):
Looks good. A few minor questions + please revert submodule change.

fritzo committed:
Revert changes to pybind11 submodule

fritzo committed:
Convert line endings to unix format

fritzo committed:
Address review comments

fritzo reviewed (commented):
Thanks for the detailed review, Adam!

fritzo committed:
Add ComposeTransform and identity_transform

fritzo committed:
Add more documentation

fritzo committed:
Fix accidental import

fritzo committed:
Fix typos in documentation

apaszke commented on torch/distributions/transforms.py:
```diff
@@ -0,0 +1,422 @@
+        else:
+            raise NotImplementedError('cache_size must be 0 or 1')
+
+    @lazy_property
```
I don't think we want to make it a lazy property, or we'll end up with reference cycles. Creation of a new inverse should be very cheap, so let's just create a new one every time

fritzo replied:
Are you sure? That will break the nice property that `t.inv.inv is t`?

fritzo replied:
Done. At least we have `t.inv.inv == t`.

apaszke replied:
Why would it break that property? We only ensure that inverse transforms have reference to the base one (and return it when asked for their inverse), while the base transform never holds any inverses. So `t.inv.inv is t`, it's just no longer true that `t.inv is t.inv` (not the same object by identity, but I'd say that comparing transforms by identity isn't a very good idea either).

apaszke replied:
BTW we could keep `t.inv.inv is t` if only we kept a weakref to the inverse. Should be quite simple:
```python
@property
def inv(self):
  maybe_inv = self._inv()
  if maybe_inv is not None:
    return maybe_inv
  inv = InverseTransform(self)
  self._inv = weakref.ref(inv)
  return inv
```

fritzo replied:
The property no longer holds for `InverseTransform`s:
```py
t = ExpTransform()
assert t.inv.inv is t      # <---- this works fine, as you observe
s = t.inv
assert s.inv.inv is not s  # <---- but this makes a second .inv copy
assert s.inv.inv == s      # <---- this is enough to make me happy :-)
```

fritzo replied:
> weak ref

Hmm so `InverseTransform`s would have strong references to their "forward" transform, but forward transforms would have only weak references to inverse transforms? That sounds reasonable. Let me see if I can get that to work using your snippet.

fritzo replied:
Thanks, @apaszke that ended up working well :smile:

apaszke reviewed (commented):
Looks good, but let's fix reference cycles.

fritzo committed:
Avoid @lazy_property in Transform.inv

fritzo commented on torch/distributions/transforms.py:
```diff
@@ -0,0 +1,422 @@
+            pass  # default behavior
+        elif cache_size == 1:
+            self._cached_x_y = None, None
+            self.__call__ = self._cached_call
```
Hmm it looks like this is another lazy reference.

fritzo replied:
Fixed.

fritzo committed:
Eliminate reference cycle doe to bound methods

fritzo committed:
Fix docs to satisfy new t.inv.inv == t equation

apaszke commented on torch/distributions/transforms.py:
```diff
@@ -146,30 +146,34 @@ class _InverseTransform(Transform):
+
+    @property
+    def inv(self):
+        return self._inv
```
I think it could have stayed as `self.inv`

fritzo replied:
That gave me a "cannot set attribute" error somewhere. I only renamed it to get tests to pass.

fritzo replied:
```
    def __init__(self, transform):
        super(_InverseTransform, self).__init__()
>       self.inv = transform
E       AttributeError: can't set attribute
```

apaszke commented on torch/distributions/transforms.py:
```diff
@@ -90,30 +88,24 @@ def __call__(self, x):
+        try:
+            x_old, y_old = self._cached_x_y
+        except AttributeError:
+            return self._call(x)
```
This will be very slow if you don't use caches. Maybe just add a flag and use an `if`?

fritzo committed:
Use weakrefs to achieve t.inv.inv is t

apaszke commented on torch/distributions/transforms.py:
```diff
@@ -0,0 +1,435 @@
+            return y_old
+        y = self._call(x)
+        self._cached_x_y = x, y
+        return y
```
My comment regarding always trying to read `_cached_x_y` still applies. I don't think it's good practice to do try catch even though you really expect the code to fail

fritzo replied:
Oh I must have missed that comment. I'll see if I can avoid try-except. Thanks for all the feedback!

fritzo replied:
What's your preferred pattern here? `if hastattr`? I try to avoid `hasattr`.

fritzo replied:
Done.

apaszke replied:
Yeah, using a flag is usually much better

fritzo committed:
Avoid try-except-AttributeError pattern

fritzo commented:
Thanks for your patience in reviewing, @apaszke!

apaszke commented:
Happy to help!

apaszke commented:
BTW should I squash this, or do you guys want to recreate some meaningful history by manual rebasing (squashed commit can only have a single author).

fritzo commented:
Squash!

apaszke commented:
Thanks @fritzo @alicanb!!
